firewood/src/merkle/stream.rs

```diff
@@ -394,22 +559,216 @@ mod tests {
+            .unwrap();
+
+        children.clone().for_each(|child_path| {
+            let key = vec![sibling_path, child_path];
+
+            merkle.insert(&key, key.clone(), root).unwrap();
+        });
 
-        let done = stream.next().await;
+        let mut keys: Vec<_> = children
+            .map(|child_path| {
+                let key = vec![branch_path, child_path];
+
+                merkle.insert(&key, key.clone(), root).unwrap();
+
+                key
+            })
+            .chain(Some(branch_key.clone()))
+            .collect();
+
+        keys.sort();
+
+        let start = keys.iter().position(|key| key == &branch_key).unwrap();
+        let keys = &keys[start..];
+
+        let mut stream = merkle.iter_from(root, branch_key.into_boxed_slice());
+
+        for key in keys {
+            let next = stream.next().await.unwrap().unwrap();
+
+            assert_eq!(&*next.0, &*next.1);
+            assert_eq!(&*next.0, key);
+        }
+
+        check_stream_is_done(stream).await;
+    }
+
+    #[tokio::test]
+    async fn start_at_key_on_extension() {
+        let missing = 0x0a;
+        let children = (0..=0x0f).filter(|x| *x != missing);
+        let mut merkle = create_test_merkle();
+        let root = merkle.init_root().unwrap();
+
+        let keys: Vec<_> = children
+            .map(|child_path| {
+                let key = vec![child_path];
+
+                merkle.insert(&key, key.clone(), root).unwrap();
+
+                key
+            })
+            .collect();
+
+        let keys = &keys[(missing as usize)..];
+
+        let mut stream = merkle.iter_from(root, vec![missing].into_boxed_slice());
+
+        for key in keys {
+            let next = stream.next().await.unwrap().unwrap();
+
+            assert_eq!(&*next.0, &*next.1);
+            assert_eq!(&*next.0, key);
+        }
+
+        check_stream_is_done(stream).await;
+    }
+
+    #[tokio::test]
+    async fn start_at_key_between_siblings() {
+        let missing = 0xaa;
+        let children = (0..=0xf)
+            .map(|val| (val << 4) + val) // 0x00, 0x11, ... 0xff
+            .filter(|x| *x != missing);
+        let mut merkle = create_test_merkle();
+        let root = merkle.init_root().unwrap();
+
+        let keys: Vec<_> = children
+            .map(|child_path| {
+                let key = vec![child_path];
+
+                merkle.insert(&key, key.clone(), root).unwrap();
+
+                key
+            })
+            .collect();
+
+        let keys = &keys[((missing >> 4) as usize)..];
+
+        let mut stream = merkle.iter_from(root, vec![missing].into_boxed_slice());
+
+        for key in keys {
+            let next = stream.next().await.unwrap().unwrap();
+
+            assert_eq!(&*next.0, &*next.1);
+            assert_eq!(&*next.0, key);
+        }
+
+        check_stream_is_done(stream).await;
+    }
+
+    #[tokio::test]
+    async fn start_at_key_greater_than_all_others() {
```

Comment:
thanks a lot for the thorough test lists! On top of that may be `start_at_key_smaller_than_all_others` ? I guess this may be already covered, but just to make it explicitly covered.